experimental/oauther/oauth2.go

```diff
@@ -0,0 +1,138 @@
+
+	"github.com/mattermost/mattermost-plugin-api/experimental/bot/logger"
+	"github.com/mattermost/mattermost-plugin-api/experimental/common"
+)
+
+const (
+	// DefaultStorePrefix is the prefix used when storing information in the KVStore by default.
+	DefaultStorePrefix = "oauth_"
+	// DefaultOAuthURL is the url the OAuther will use to register its endpoints by default.
+	DefaultOAuthURL = "/oauth2"
+	// DefaultConnectedString is the string shown to the user when the oauth flow is completed by default.
+	DefaultConnectedString = "Successfully connected. Please close this window."
+	// DefaultOAuth2StateTimeToLive is the duration the states from the OAuth flow will live in the KVStore by default.
+	DefaultOAuth2StateTimeToLive = 5 * time.Minute
+)
+
+const (
+	connectURL  = "/connect"
+	completeURL = "/complete"
+)
+
+// OAuther defines an object able to perform the OAuth flow.
+type OAuther interface {
+	// GetToken returns the oauth token for userID, or error if it does not exist or there is any store error.
+	GetToken(userID string) (*oauth2.Token, error)
+	// GetConnectURL returns the URL to reach in order to start the OAuth flow.
+	GetConnectURL() string
+	// Deauthorize removes the token for userID. Return error if there is any store error.
+	Deauthorize(userID string) error
+	// ServeHTTP implements http.Handler
+	ServeHTTP(w http.ResponseWriter, r *http.Request)
+}
+
+type oAuther struct {
+	pluginURL             string
+	config                *oauth2.Config
+	onConnect             func(userID string, token *oauth2.Token)
+	store                 common.KVStore
+	logger                logger.Logger
+	storePrefix           string
+	oAuthURL              string
+	connectedString       string
+	oAuth2StateTimeToLive time.Duration
+}
+
+/*NewOAuther creates a new OAuther.
+
+- pluginURL: The base url for the plugin (e.g. instance.com/plugins/pluginid).
+
+- oAuthConfig: The configuration of the Authorization flow to perform.
+
+- onConnect: What to do when the Authorization process is complete.
+
+- store: A KVStore to store the data of the OAuther.
+
+- l Logger: A logger to log errors during authorization.
+
+- options: Optional options for the OAuther. Available options are StorePrefix, OAuthURL, ConnectedString and OAuth2StateTimeToLive.
+*/
+func NewOAuther(
+	pluginURL string,
+	oAuthConfig *oauth2.Config,
+	onConnect func(userID string, token *oauth2.Token),
+	store common.KVStore,
+	l logger.Logger,
+	options ...Option,
+) OAuther {
+	o := &oAuther{
+		pluginURL:             pluginURL,
+		config:                oAuthConfig,
+		onConnect:             onConnect,
+		store:                 store,
+		logger:                l,
+		storePrefix:           DefaultStorePrefix,
+		oAuthURL:              DefaultOAuthURL,
+		connectedString:       DefaultConnectedString,
+		oAuth2StateTimeToLive: DefaultOAuth2StateTimeToLive,
+	}
+
+	for _, option := range options {
+		option(o)
+	}
+
+	o.config.RedirectURL = o.pluginURL + o.oAuthURL + "/complete"
```

Comment:
E.g. GitHub doesn't allow OAuth Applications to set the `RedirectURL` itself. Admins have to set it via a UI. Is this the same for other OAuth applications?

Reply:
As far as I know, you set the RedirectURL through the UI, and the configuration must match the RedirectURL. If the configuration and what is set on the UI mismatch, you will receive an error message during the OAuth flow saying that the redirect URL does not match.

Is anyhow different in Github?

Reply:
AFAIK GitHub only uses the URL from the UI. 

The reasoning I'm bringing this question up is that it's weird to require a `pluginURL` from a library that should help you setup an URL. Any ideas on how to drop that parameter?

Reply:
It is complicated to drop it. We need the plugin URL to also get the connect URL. We could require the whole API, and look for the URL ourselves, but I prefer to ask for "just what is needed".

Regarding the RedirectURL, I wonder if passing it to Github will pose any problem. If it does, then we would have to add some option to remove or to add it.

Reply:
Could we add a second constructor, `NewFromClient(pluginapi *pluginapi.Client, ...)` that streamlines this?

Reply:
In my opinion, using the pluginapi client have not solved many pain points, but I added it.

Right now, you can omit the store (you can take it from the plugin api) and you don't have to set the URL, but it still needs the plugin id (AFAIK it is only on the manifest, and pluginapi does not have access to the manifest).

We could consider creating also a default logger, but since there is more configuration to be handled on the logger, I think it would be best to leave the logger as it is.